File: ios/Tuner/Metronome/MetronomeViewModel.swift
import Combine
import Foundation

extension TickAccent {
    var persistName: String {
        switch self {
        case .accent: return "accent"
        case .normal: return "normal"
        case .muted: return "muted"
        }
    }

    static func from(persistName: String) -> TickAccent {
        switch persistName {
        case "accent": return .accent
        case "muted": return .muted
        default: return .normal
        }
    }
}

/// 节拍器面板 ViewModel（与 Android MetronomeViewModel 同构；节奏调度全在 core）。
@MainActor
final class MetronomeViewModel: ObservableObject {
    static let bpmMin = 30.0
    static let bpmMax = 250.0

    @Published var bpm: Double = 120.0 {
        didSet { apply { $0.setBpm(bpm: bpm) }; save() }
    }
    @Published var beatsPerBar: Int = 4 { didSet { save() } }
    @Published var beatUnit: Int = 4 { didSet { save() } }
    @Published var accents: [TickAccent] = MetronomeViewModel.defaultAccents(4) {
        didSet { apply { $0.setAccents(accents: accents) }; save() }
    }
    @Published var accentSound: TickSoundKind = .bell { didSet { applySounds(); save() } }
    @Published var normalSound: TickSoundKind = .click { didSet { applySounds(); save() } }
    @Published private(set) var playing = false
    @Published private(set) var currentBeat = -1
    @Published private(set) var flashSeq: Int = 0

    private let factory: MetronomeEngineFactory
    private let player: MetronomePlayer
    private let sounds: [TickSoundKind: [Float]]
    private let clock: () -> UInt64
    private var engine: MetronomeEngine?
    private var cancellables = Set<AnyCancellable>()

    init(
        factory: MetronomeEngineFactory = UniffiFactories.metronome,
        player: MetronomePlayer = MetronomePlayer(),
        sounds: [TickSoundKind: [Float]] = TickSounds.buildAll(),
        clock: @escaping () -> UInt64 = { UInt64(ProcessInfo.processInfo.systemUptime * 1000) }
    ) {
        self.factory = factory
        self.player = player
        self.sounds = sounds
        self.clock = clock
        restore()
        engine = factory.create(config: currentConfig())
        applySounds()

        player.ticks
            .receive(on: DispatchQueue.main)
            .sink { [weak self] ev in
                guard let self else { return }
                let waitMs = Int64(ev.atMs) - Int64(self.clock())
                if waitMs > 0 {
                    Task { [weak self] in
                        try? await Task.sleep(nanoseconds: UInt64(waitMs) * 1_000_000)
                        self?.currentBeat = ev.beatIndex
                        self?.flashSeq += 1
                    }
                } else {
                    self.currentBeat = ev.beatIndex
                    self.flashSeq += 1
                }
            }
            .store(in: &cancellables)
    }

    static func defaultAccents(_ beats: Int) -> [TickAccent] {
        (0..<beats).map { $0 == 0 ? TickAccent.accent : TickAccent.normal }
    }

    static let commonTimeSignatures: [(Int, Int)] = [
        (1, 4), (2, 4), (3, 4), (4, 4), (5, 4), (6, 4),
        (3, 8), (6, 8), (9, 8), (12, 8),
    ]

    private func currentConfig() -> MetronomeConfig {
        MetronomeConfig(
            sampleRate: TickSounds.sampleRate,
            bpm: bpm,
            beatsPerBar: UInt8(beatsPerBar),
            beatUnit: UInt8(beatUnit),
            accents: accents
        )
    }

    private func apply(_ f: (MetronomeEngine) -> Void) {
        engine.map(f)
    }

    private func applySounds() {
        apply {
            $0.setClickSamples(
                accent: sounds[accentSound] ?? TickSounds.synthesize(accentSound),
                normal: sounds[normalSound] ?? TickSounds.synthesize(normalSound)
            )
        }
    }

    func setBpm(_ v: Double) {
        bpm = v.clamped(to: Self.bpmMin...Self.bpmMax)
    }

    func adjustBpm(_ delta: Double) { setBpm(bpm + delta) }

    func tap() {
        let tsSamples = UInt64(Double(clock()) * TickSounds.sampleRate / 1000.0)
        engine.map { setBpm($0.tap(timestampSamples: tsSamples)) }
    }

    func setTimeSignature(beats: Int, unit: Int) {
        beatsPerBar = beats
        beatUnit = unit
        accents = Self.defaultAccents(beats)
        apply { $0.setTimeSignature(beats: UInt8(beats), unit: UInt8(unit)) }
    }

    func cycleAccent(_ index: Int) {
        guard accents.indices.contains(index) else { return }
        accents[index] = switch accents[index] {
        case .accent: .normal
        case .normal: .muted
        case .muted: .accent
        }
    }

    func togglePlay() {
        if playing { pause() } else { play() }
    }

    func play() {
        guard !playing else { return }
        engine.map { $0.start(atSample: 0) }
        engine.map { player.start(engine: $0) }
        playing = true
    }

    func pause() {
        guard playing else { return }
        player.stop()
        engine.map { $0.stop() }
        playing = false
        currentBeat = -1
    }

    // MARK: 持久化（UserDefaults）

    private func save() {
        let d = UserDefaults.standard
        d.set(bpm, forKey: "metro_bpm")
        d.set(beatsPerBar, forKey: "metro_beats")
        d.set(beatUnit, forKey: "metro_unit")
        d.set(accents.map(\.persistName), forKey: "metro_accents")
        d.set(accentSound.rawValue, forKey: "metro_accent_sound")
        d.set(normalSound.rawValue, forKey: "metro_normal_sound")
    }

    private func restore() {
        let d = UserDefaults.standard
        bpm = (d.object(forKey: "metro_bpm") as? Double ?? 120.0).clamped(to: Self.bpmMin...Self.bpmMax)
        beatsPerBar = d.object(forKey: "metro_beats") as? Int ?? 4
        beatUnit = d.object(forKey: "metro_unit") as? Int ?? 4
        if let raw = d.array(forKey: "metro_accents") as? [String] {
            let a = raw.map(TickAccent.from(persistName:))
            if a.count == beatsPerBar { accents = a }
        }
        accentSound = TickSoundKind(rawValue: d.string(forKey: "metro_accent_sound") ?? "") ?? .bell
        normalSound = TickSoundKind(rawValue: d.string(forKey: "metro_normal_sound") ?? "") ?? .click
    }
}

File: android/app/src/main/kotlin/com/liuchong/tunar/ui/metronome/MetronomeViewModel.kt
package com.liuchong.tunar.ui.metronome

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.liuchong.tunar.audio.MetronomePlayer
import com.liuchong.tunar.audio.TickEvent
import com.liuchong.tunar.corebinding.MetronomeEngine
import com.liuchong.tunar.corebinding.MetronomeEngineFactory
import com.liuchong.tunar.corebinding.TickSoundKind
import com.liuchong.tunar.corebinding.TickSounds
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.tunar_core.MetronomeConfig
import uniffi.tunar_core.TickAccent

/** 节拍器面板 UI 状态。 */
data class MetronomeUiState(
    val bpm: Double = 120.0,
    val beatsPerBar: Int = 4,
    val beatUnit: Int = 4,
    val accents: List<TickAccent> = defaultAccents(4),
    val accentSound: TickSoundKind = TickSoundKind.BELL,
    val normalSound: TickSoundKind = TickSoundKind.CLICK,
    val playing: Boolean = false,
    /** 当前闪拍（-1 = 无）；flashSeq 用于触发一次性动画。 */
    val currentBeat: Int = -1,
    val flashSeq: Long = 0L,
)

/** 默认重音型：首拍重拍，其余弱拍。 */
fun defaultAccents(beats: Int): List<TickAccent> =
    List(beats) { if (it == 0) TickAccent.ACCENT else TickAccent.NORMAL }

/** 常用拍号集（spec-ui §3：1/4–12/8）。 */
val COMMON_TIME_SIGNATURES: List<Pair<Int, Int>> = listOf(
    1 to 4, 2 to 4, 3 to 4, 4 to 4, 5 to 4, 6 to 4,
    3 to 8, 6 to 8, 9 to 8, 12 to 8,
)

/**
 * 节拍器面板 ViewModel。节奏调度全在 Rust core 引擎；
 * 本类只做参数映射、选择状态与闪拍事件转发。
 */
class MetronomeViewModel(
    private val engineFactory: MetronomeEngineFactory,
    private val player: MetronomePlayer,
    private val sounds: Map<TickSoundKind, List<Float>>,
    private val savedState: SavedStateHandle,
    private val clock: () -> Long = { android.os.SystemClock.uptimeMillis() },
) : ViewModel() {

    private val _uiState = MutableStateFlow(restoreState())
    val uiState: StateFlow<MetronomeUiState> = _uiState.asStateFlow()

    private var engine: MetronomeEngine? = null

    init {
        // 引擎常驻（tap tempo 在未播放时也要工作）
        engine = engineFactory.create(currentConfig()).also { e ->
            applySounds(e, _uiState.value.accentSound, _uiState.value.normalSound)
        }
        // 闪拍事件：延时到呈现时刻再更新状态（与声音同步）
        viewModelScope.launch {
            player.ticks.collect { ev: TickEvent? ->
                if (ev != null) {
                    val wait = ev.atMs - clock()
                    if (wait > 0) delay(wait)
                    _uiState.update {
                        it.copy(currentBeat = ev.beatIndex, flashSeq = it.flashSeq + 1)
                    }
                }
            }
        }
    }

    private fun restoreState(): MetronomeUiState {
        val beats = savedState.get<Int>(KEY_BEATS) ?: 4
        val accents = savedState.get<Array<String>>(KEY_ACCENTS)
            ?.map { TickAccent.valueOf(it) }
            ?.takeIf { it.size == beats }
            ?: defaultAccents(beats)
        return MetronomeUiState(
            bpm = (savedState.get<Double>(KEY_BPM) ?: 120.0).coerceIn(BPM_MIN, BPM_MAX),
            beatsPerBar = beats,
            beatUnit = savedState.get<Int>(KEY_UNIT) ?: 4,
            accents = accents,
            accentSound = restoreSound(KEY_ACCENT_SOUND, TickSoundKind.BELL),
            normalSound = restoreSound(KEY_NORMAL_SOUND, TickSoundKind.CLICK),
        )
    }

    private fun restoreSound(key: String, fallback: TickSoundKind): TickSoundKind {
        val savedName = savedState.get<String>(key) ?: return fallback
        return TickSoundKind.entries.firstOrNull { it.name == savedName } ?: fallback
    }

    private fun currentConfig() = _uiState.value.let {
        MetronomeConfig(
            sampleRate = TickSounds.SAMPLE_RATE,
            bpm = it.bpm,
            beatsPerBar = it.beatsPerBar.toUByte(),
            beatUnit = it.beatUnit.toUByte(),
            accents = it.accents,
        )
    }

    /** 设置 BPM（钳制 30–250），引擎即时生效。 */
    fun setBpm(bpm: Double) {
        val clamped = bpm.coerceIn(BPM_MIN, BPM_MAX)
        savedState[KEY_BPM] = clamped
        _uiState.update { it.copy(bpm = clamped) }
        engine?.setBpm(clamped)
    }

    /** 步进调整。 */
    fun adjustBpm(delta: Int) = setBpm(_uiState.value.bpm + delta)

    /** tap tempo（≥2 次生效；core 取最近 ≤4 间隔中位数）。 */
    fun tap() {
        val tsSamples = (clock() * TickSounds.SAMPLE_RATE / 1000.0).toLong()
        engine?.let { e -> setBpm(e.tap(tsSamples)) }
    }

    /** 设置拍号；重音型重置为默认（首拍重拍）。 */
    fun setTimeSignature(beats: Int, unit: Int) {
        val accents = defaultAccents(beats)
        savedState[KEY_BEATS] = beats
        savedState[KEY_UNIT] = unit
        savedState[KEY_ACCENTS] = accents.map { it.name }.toTypedArray()
        _uiState.update { it.copy(beatsPerBar = beats, beatUnit = unit, accents = accents) }
        engine?.setTimeSignature(beats, unit)
        engine?.setAccents(accents)
    }

    /** 点击某拍圆点：重拍 → 普通 → 静音 循环。 */
    fun cycleAccent(index: Int) {
        val current = _uiState.value.accents
        if (index !in current.indices) return
        val next = current.mapIndexed { i, a ->
            if (i == index) {
                when (a) {
                    TickAccent.ACCENT -> TickAccent.NORMAL
                    TickAccent.NORMAL -> TickAccent.MUTED
                    TickAccent.MUTED -> TickAccent.ACCENT
                }
            } else {
                a
            }
        }
        savedState[KEY_ACCENTS] = next.map { it.name }.toTypedArray()
        _uiState.update { it.copy(accents = next) }
        engine?.setAccents(next)
    }

    /** 重拍音色。 */
    fun setAccentSound(kind: TickSoundKind) {
        savedState[KEY_ACCENT_SOUND] = kind.name
        _uiState.update { it.copy(accentSound = kind) }
        engine?.let { e -> applySounds(e, kind, _uiState.value.normalSound) }
    }

    /** 弱拍音色。 */
    fun setNormalSound(kind: TickSoundKind) {
        savedState[KEY_NORMAL_SOUND] = kind.name
        _uiState.update { it.copy(normalSound = kind) }
        engine?.let { e -> applySounds(e, _uiState.value.accentSound, kind) }
    }

    private fun applySounds(e: MetronomeEngine, accent: TickSoundKind, normal: TickSoundKind) {
        val a = sounds[accent] ?: TickSounds.synthesize(accent)
        val n = sounds[normal] ?: TickSounds.synthesize(normal)
        e.setClickSamples(a, n)
    }

    /** 播放/停止切换。返回 true 表示开始播放（供 UI 启动保活 Service）。 */
    fun togglePlay(): Boolean = if (_uiState.value.playing) {
        pause()
        false
    } else {
        play()
        true
    }

    fun play() {
        if (_uiState.value.playing) return
        engine?.start(0)
        engine?.let { player.startLoop(it) }
        _uiState.update { it.copy(playing = true, currentBeat = -1) }
    }

    fun pause() {
        if (!_uiState.value.playing) return
        player.stopLoop()
        engine?.stop()
        _uiState.update { it.copy(playing = false, currentBeat = -1) }
    }

    override fun onCleared() {
        player.stopLoop()
        engine?.close()
        engine = null
    }

    companion object {
        const val BPM_MIN = 30.0
        const val BPM_MAX = 250.0
        private const val KEY_BPM = "bpm"
        private const val KEY_BEATS = "beats"
        private const val KEY_UNIT = "unit"
        private const val KEY_ACCENTS = "accents"
        private const val KEY_ACCENT_SOUND = "accentSound"
        private const val KEY_NORMAL_SOUND = "normalSound"
    }
}
